src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ILEmit/ILEmitResolverBuilder.cs

```diff
@@ -309,8 +326,36 @@ private ILEmitResolverBuilderRuntimeContext GenerateMethodBody(ServiceCallSite c
 
                 Label skipCreationLabel = context.Generator.DefineLabel();
                 Label returnLabel = context.Generator.DefineLabel();
+                Label defaultLabel = context.Generator.DefineLabel();
+
+                // Check if scope IsRootScope
+                context.Generator.Emit(OpCodes.Nop);
+                context.Generator.Emit(OpCodes.Ldarg_1);
+                context.Generator.Emit(OpCodes.Callvirt, ScopeIsRootScope);
+                context.Generator.Emit(OpCodes.Stloc_3);
+                context.Generator.Emit(OpCodes.Ldloc_3);
+                context.Generator.Emit(OpCodes.Brfalse_S, defaultLabel);
+                context.Generator.Emit(OpCodes.Nop);
+
+                // Return resolved value from CallSiteRuntimeResolver for root scope
+                if (true || CallSiteGetter != null)
+                {
+                    // TODO - find out what in else condition is broken
+                    context.Generator.Emit(OpCodes.Nop); 
+                }
+                else
+                {
+                    context.Generator.Emit(OpCodes.Call, CallSiteRuntimeResolverInstanceField);
+                    context.Generator.Emit(OpCodes.Ldarg_0);
+                    context.Generator.Emit(OpCodes.Callvirt, CallSiteGetter);
+                    context.Generator.Emit(OpCodes.Ldarg_1);
+                    context.Generator.Emit(OpCodes.Callvirt, CallSiteRuntimeResolverResolveMethod);
+                    context.Generator.Emit(OpCodes.Stloc_S, 4);
```

Comment:
we should use a LocalBuilder for local variables instead of indexes, here and elsewhere